### PR TITLE
feat(component): change updating cell strategy for text field

### PR DESCRIPTION
**Pull request by bc-apostoliuk**
## What?
Change `Worksheet` updating cell strategy for the `text` and `number` fields.

- Use cases:

1. When opening the Inventory Editor the top left cell gets the Focused state (so merchants can enter value right after accessing the page with no additional clicks on the cell.

2. When a cell is Focused (by keyboard or mouse) it can be edited straight away by:

- Inserting char or numeric values.

- Press Enter from the keyboard.

- Single mouse click.

3. When a cell is Edited all changes can be saved by the “lost focus” user’s action:

- pressing the Tab button - the cell to the right becomes Focused.

- pressing Enter button - the cell to the bottom becomes Focused.

- selecting any other cell by mouse - selected cell becomes Focused.

4. When a cell is Edited all changes can be canceled by pressing the Esc button. Edited cell changes its state to the Focused one.

5. When changing the value within the cell, the previous one should be replaced with the newly entered value (currently new value is added to the existing one).

## Why?

It needs to be changed to the new strategy according to the new guideline.

## Screenshots/Screen Recordings

https://user-images.githubusercontent.com/84462142/192572546-3300a55b-8dab-4772-8764-366451240372.mov

## Testing/Proof

Explain how you tested your new feature or bugfix. Please refrain from one-liners like, `Added tests.`; expand upon how you tested your changes like, `Added local tests to the new component and ensured test coverage was met. Manually tested on Safari@x.x.x.`.

### CODE REVIEW

**Review by vdm-kr — APPROVED** [review]
lgtm

**Review by jorgemoya — CHANGES_REQUESTED** [review]
Will need to figure out how we want to handle arrow navigation with this behavior. Do we just lose it? Do we need to fix it? Tests are breaking because of this.

**Comment by chanceaclark** [comment]
IMO, we probably shouldn't introduce this behavior. Merchants/users already have the habit of pressing enter to edit a cell based on using Excel or Google Sheets. And like Jorge mentioned, there are other behaviors we need to account for including event handling (onChange, onError, etc...). We also need to add a few tests for this.

**Comment by vdm-kr** [comment]
@chanceaclark actually you don't have to press enter to edit a cell in excel/google docs or our old balk edit. And that's what we wanted to achieve

**Comment by chanceaclark** [comment]
> @chanceaclark actually you don't have to press enter to edit a cell in excel/google docs or our old balk edit. And that's what we wanted to achieve

Ah I see now... however I don't think that this change is the _right fix_ for the problem we are trying to solve. I know you can start typing when you are focused on a cell, but you need to initially be focused on that cell so this onblur/onfocus method is not the right approach.

https://user-images.githubusercontent.com/10539418/192005333-64ab5639-57fc-45b7-a52a-4e3e4563a2ed.mov

The fix I see is that if you are focused on a cell and use a keyboard to start typing, then you start editing – not click on a cell and now it's in editing mode.

**Comment by DmytroSirenko** [comment]
@kristinapototska, FYI.

**Comment by kristinapototska** [comment]
> @kristinapototska, FYI. There is a discussion about saving values in cells on the "lost focus" behavior.

Thanks Dmytro. I can see that @vdm-kr has already shared the expected behaviour here. We do want the experience to be intuitive and similar to worksheet editing experience available with other tools, like good sheets etc.
Also we can use bulk edit as an example.
When the focus is on cell (for text fields) you can start typing new value right at once with no need to do additional clicks. And we've previously discussed with Vadym that when inventory editor is opened, ideally would have the default focus on `stock` cell so that merchants can start typing the value right after the page is loaded which would save time and reduce the amount of clicks. 
If the `focus is lost` (e.g. merchant clicked outside the table, the editing will be unavailable till merchant click on the cell again for it to be in focus.

What's the open question that we still need to figure out here?

**Comment by bc-apostoliuk** [comment]
@chanceaclark, @jorgemoya thanks for the comments, I have changed the implementation approach, please review again.

**Comment by jorgemoya** [comment]
> @chanceaclark, @jorgemoya thanks for the comments, I have changed the implementation approach, please review again.

Can you share a recording of the new approach?

**Comment by bc-apostoliuk** [comment]
@jorgemoya I`ve reattached the new recording to the PR. Please check it there.

File: packages/big-design/src/components/Worksheet/Cell/Cell.tsx [hunk 1]
```diff
@@ -158,18 +164,19 @@ const InternalCell = <T extends WorksheetItem>({
...
-    isEdited,
-    isEditing,
     options,
+    formatting,
     rowId,
+    disabled,
+    editWithValue,
+    isEdited,
+    handleKeyDown,
     renderedValue,
```
Comment (jorgemoya) [inline packages/big-design/src/components/Worksheet/Cell/Cell.tsx:179]: We prefer to `abc`.

File: packages/big-design/src/components/Worksheet/hooks/useEditableCell/useEditableCell.ts [hunk 2]
```diff
@@ -40,9 +41,22 @@ export const useEditableCell = <T extends WorksheetItem>(cell: Cell<T>) => {
...
+        updateItems(
+          [selectedCells[0]],
+          [isNumberCell ? Number(event?.target.value) : event?.target.value],
+        );
+      }
+
+      restoreFocus();
+    },
+    [restoreFocus, selectedCells, updateItems],
+  );
```
Comment (jorgemoya) [inline packages/big-design/src/components/Worksheet/hooks/useEditableCell/useEditableCell.ts:59]: Hmm if we had a multiple selection of cells (in the future), how would this be managed? Can we pass the the `cell` as a param instead?

Comment (bc-apostoliuk) [reply packages/big-design/src/components/Worksheet/hooks/useEditableCell/useEditableCell.ts:59]: @jorgemoya This behavior should work only for the first selected cell. What do you mean by passing the `cell` as a param?

File: packages/big-design/src/components/Worksheet/Cell/Cell.tsx (hunk 1 above)
Comment (bc-apostoliuk) [reply packages/big-design/src/components/Worksheet/Cell/Cell.tsx]: @jorgemoya that was the linter auto-fix, I have returned it back))

**Review by jorgemoya — APPROVED** [review]
Looks good to me 👍

**Review by MariaJose — APPROVED** [review]

**Comment by eugene-polev** [comment]
@bc-apostoliuk , @vdm-kr should we highlight the value inside the cell when it is focused as we do for inputs?

Screenshot 2022-09-28 at 12 26 06
Screenshot 2022-09-28 at 12 24 10

**Comment by bc-apostoliuk** [comment]
@eugene-polev I guess not if we are trying to reproduce the Excel behavior, since over there the selection of the field does not highlight the text inside.

**Pull request merged**